### ISSUE TITLE
SMPTE 240M discussion

### ISSUE THREAD

**Issue by sekrit-twc** [sekrit-twc/zimg#155]

**Comment by dwbuiten** [comment]
I would like to note NLEs and QuickTime both match the reverted zimg behavior on ctual 240M files (and I have shared at last one, but I have many). There was a huge uptick in complains from users when the original commit went into prod. Michael Bradshaw / @mjbshaw (from YouTube) on that thread probably experience similar.

You can maybe argue the behavior pre-revert was "most correct" based on docs, but it does not match how any software or files act in the real world, and thus is defacto wrong, in my opinion. It is not useful to say "but this is technically correct" if it doesn't match anything else in the real world. There is plenty of evidence in how standard NLEs and players render 240M, an how they create files. I have thus far seen zero evidence that the pre-revert behavior was correct, other than arguing technical semantics based on docs.

**Comment by ValZapod** [comment]
> NLEs and QuickTime both match

QuickTime does not even support BT.1886, so who cares what it does (it uses gamma 1.961). Also it has TrueTone so the test is also quite difficult to do.

As for NLEs (I doubt any NLE supports this old and deprecated transfer) the mpv is only one that can support it and ffplay (no need to include my fix https://github.com/FFmpeg/FFmpeg/commit/3249b963421cf5ff0e6adf2e38c86194fdf50929) with this: ffplay C:\Users\ZAQU\2.mp4 -vf scale=in_color_matrix=smpte240m,format=gbrp or just auto instead of smpte240m if tagged (that is only colorspace, transfer and primaries both affect the picture too and is not supported by ffplay, only by mpv).

> There was a huge uptick in complains

Why those users even use that transfer?? It is deprecated in favour of BT.709 OETF which is using BT.1886 EOTF, which is almost sRGB for LCD displays and 2.4 pure gamma for OLED. Also, Youtube does not support this transfer, AFAIK, and since most use Chromium/Chrome to watch it, as a dev of Chrome, I can say it forces sRGB EOTF for BT.709, so there is no point in arguing about this, since everything else on the web is targeting sRGB and color management of WCG images is very broken (see: https://bugs.chromium.org/p/chromium/issues/detail?id=820233) and CSS WCG support is still a draft W3C draft.

> commit went into prod.

BTW, why is Youtube using zimg? Also, not like Youtube even supports sYCC and stuff.

> I have a _lot_ of files from a lot of software

Please give at least mediainfo of those files. Does it use ffmpeg and/or zimg? Then it is not surprising, since ffplay and mxf encoder in ffmpeg and ffmpeg documentation had this bug in DOCS only that I fixed: https://github.com/FFmpeg/FFmpeg/commit/6b1268f8c371c2e6efa5c8465b85ae9025254184 but it did propagate elsewhere: https://github.com/FFmpeg/FFmpeg/commit/91ba771a3fee9757273416789ba01e1a7c62890b https://github.com/FFmpeg/FFmpeg/commit/3249b963421cf5ff0e6adf2e38c86194fdf50929

THE FACT OF THE MATTER is that SMPTE240M EOTF and OETF are not the same as in any other. This is the fact. See ffmpeg source code: https://github.com/FFmpeg/FFmpeg/blob/870bfe16a12bf09dca3a4ae27ef6f81a2de80c40/libavutil/color_utils.c#L83

Also, the fact is H.273 does use different transfer for 240M. There will be always broken files, for example of some 8K satellites I saw BT.2020-CL matrix which should have been BT.2020-NCL. That is just the way it is. If google wants it can support some in tree patches for this.

**Comment by dwbuiten** [comment]
> QuickTime does not even support BT.1886, so who cares what it does

Almost everybody working in profesional (and amateur) video editing cares what QuickTime does.

> As for NLEs (I doubt any NLE supports this old and deprecated transfer)

You would be surprised.

> Why those users even use that transfer??

I do not read minds.

> Also, Youtube does not support this transfer,

I would be surprised if they did not support it for ingest.

> BTW, why is Youtube using zimg? Also, not like Youtube even supports sYCC and stuff.

They're not AFAIK, but the experience wrt 240m source files is relevant.

> Please give at least mediainfo of those files. Does it use ffmpeg and/or zimg

I've provided some to @sekrit-twc, but I'm not clear to provide user files in the open. I have plenty that are not made by FFmpeg or zimg.

In any case, I don't think I will be spending any more of my time Arguing On The Internet with someone who appears quite angry. If the revert is reverted, then fine, I can work around it on my end manually. I merely reported the feedback.

**Comment by ValZapod** [comment]
> video editing cares what QuickTime does

Except everybody really pro uses reference monitors with BT.1886 and special player, so false.

> I would be surprised if they did not support it for ingest.

But their apps and what else do not.

> provide user files or info in the open.

you can just give primaries/trc/range/matrix part.

**Comment by kodabb** [comment]
I don't think we should break user files for the sake of being technically correct on an old and deprecated format, as it could be argued that the the current zimg implementation (post revert) is the behavior needed by this de-facto standard.

if a stricter implementation is required maybe a custom option could be added to support this specific use case. jm2c

**Comment by ValZapod** [comment]
> custom option could be added to support

That would be nice. Though I still do not get how those users have this transfer. I never saw it!

**Comment by sekrit-twc** [comment]
My two cents: it can never be correct to apply inverse 240M OETF followed by 1886 EOTF to do a gamma conversion, because 601, 709, and all other CRT-based transfer functrions imply an end-to-end 1.2 gamma boost. It is not clear to me that 240M is different from 601 and 709, but if it were, it would still be incorrect to convert to 709 by applying 1886 on top of 240M, as that would add a second layer of gamma boost.

**Comment by ValZapod** [comment]
> inverse 240M OETF

You do not need to invert 240M OETF, you can just use the 240M EOTF, which is defined in SMPTE doc.

> imply an end-to-end 1.2 gamma boost

Yes, but that is just dark surround correction. Because most video is done in bright environment and the end result is presented in much darker environment, in case of BT.1886 2.4 gamma is for dark room and sRGB EOTF can be used as a good approximate for light room.

> convert to 709 by applying 1886 on top of 240M

Yes, it should be as before revert apply SMPTE240M EOTF, apply BT.709 OETF, that can then be presented using BT.1886.

**Comment by ValZapod** [comment]
@haasn What do you think? Should SMPTE 240M transfer be EOTF and OETF from the docs? Looks like mpv has some problems with SMPTE 240M, i.e. is just using BT.1886 for it too? See: 
![Uploading Screenshot_20210808-212225_mpv.png…]()

**Comment by sekrit-twc** [comment]
Updating this issue with some additional context.

ValZapod wrote:
>since most use Chromium/Chrome to watch it, as a dev of Chrome, I can say

Contributor A wrote:
>I missed ValZapod being a Googler.
>A nice quote from chromium's bug tracker:
>
>>If it makes you feel any better, you're absolutely correct in principle, and we're objectively wrong.  We simply made this decision by allowing more than that objective principle to affect our decision, and we know how and why we're wrong, and I think we're all still pretty much okay with that decision.
>>
>Yeah, Val is not a Project Member :)

Contributor B wrote:
>>and as a dev of Chrome, I can say
>>
>I can find no evidence of this

Contributor A wrote:
>>I can find no evidence of this
>>
>At most, he posted a  rejected patch. He does not have Project Member badge on the Chromium issue tracker

**Comment by ValZapod** [comment]
> does not have Project Member badge

That is only available for @chromium and @google email.

**Comment by ValZapod** [comment]
Oh, and BTW. These guys know much more than you do and even I. Same: https://github.com/colour-science/colour/blob/02612fef79cddbada324def766b8a93283430908/colour/models/rgb/transfer_functions/smpte_240m.py#L133

**Comment by ValZapod** [comment]
Okay I now see that in SMPTE 170M the EOTF is also just the inverse instead of being BT.1886, which was only done in 2011.

See: https://www.itu.int/dms_pubrec/itu-r/rec/bt/R-REC-BT.1700-0-200502-I!!ZPF-E.zip (s170m file inside).

Still same BT.1886 cannot be used for SMPTE240M transfer!

**Comment by mirh** [comment]
So.. I don't want to disturb you gentlemen (even because this is not my field, and I couldn't find the last 1999 spec anywhere)
Though I just wanted to point out 170M has a lot of "[ifs](https://forum.doom9.org/showthread.php?t=182665)" and "[buts](https://ieeexplore.ieee.org/document/7289811)" between what the datasheets say, and what the thing in the real world ended up being.

**Comment by ValZapod** [comment]
170M is just BT.601 matrix, but BT.601 matrix was derived from BT.470 System M analog TV. It is not interesting. 240M uses matrix derived from primaries defined in 170M, SMPTE **C**onrac.

As for your link for 170M transfer that is wrong. 170M uses (as per H.273) the very same 601/709/2020-10 transfer with BT.1886 EOTF.

**Comment by John-Greg** [comment]
Just a newbie here.  I don't know if this is helpful, interesting, or totally irrelevant:
---I was preparing to capture footage from an early 1990's Betacam SP master. As I studied about these colorimetry/SMPTE standards, there was this comment on Wikipedia (https://en.wikipedia.org/wiki/NTSC):
"Japanese NTSC never changed primaries and whitepoint to SMPTE "C", continuing to use the 1953 [_470m?_] NTSC primaries and whitepoint."--which I felt was relevant to me, since this is a Japanese Betacam master tape.

**Comment by ValZapod** [comment]
> Japanese NTSC never changed primaries and whitepoint to SMPTE "C", continuing to use the 1953 [_470m?_]

Yes NTSC 1953 is BT.470 System M, you are right! As for this issue it has nothing to do with your issue, we are talking about transfer, not primaries. For SD no one ever cared about primaries though! Only mpv player color manages 1953 to sRGB's primaries and transfer, BT.709 + sRGB transfer or to Display P3 used by Apple, or 2020 used by LG C9. Etc.

**Comment by John-Greg** [comment]
Thank you. You guys are amazing!

**Comment by ValZapod** [comment]
Okay, at the very least the change of primaries should happen using inverse OETF, because color managment on scene referred sources should happen ON SCENE LIGHT. Then you can apply BT.1886, whatever.
